lib/iris/experimental/regrid.py

```diff
@@ -827,43 +874,59 @@ def regrid_weighted_curvilinear_to_rectilinear(src_cube, weights, grid_cube):
-            sx.units != sy.units:
+    #
+    # FOR NOW: minimal hacks to allow non-latlon grids
+    #
+    if sx.units != sy.units:
         msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
-            'have units of degrees or radians.'
+            'have the same units.'
         raise ValueError(msg.format(sx.name(), sy.name()))
 
-    if tx.units.modulus is None or ty.units.modulus is None or \
-            tx.units != ty.units:
+    if tx.units != ty.units:
         msg = 'The target grid cube x ({!r}) and y ({!r}) coordinates must ' \
-            'have units of degrees or radians.'
+            'have the same units.'
         raise ValueError(msg.format(tx.name(), ty.name()))
 
-    if sx.units != tx.units:
-        msg = 'The source cube and target grid cube must have x and y ' \
-            'coordinates with the same units.'
-        raise ValueError(msg)
+#    if sx.units != tx.units:
+#        msg = 'The source cube and target grid cube must have x and y ' \
+#            'coordinates with the same units.'
+#        raise ValueError(msg)
 
     if sx.ndim != sy.ndim:
         msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
             'have the same dimensionality.'
         raise ValueError(msg.format(sx.name(), sy.name()))
 
-    if sx.ndim != 2:
-        msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
-            'be 2D auxiliary coordinates.'
-        raise ValueError(msg.format(sx.name(), sy.name()))
+#    if sx.ndim != 2:
+#        msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
+#            'be 2D auxiliary coordinates.'
+#        raise ValueError(msg.format(sx.name(), sy.name()))
 
     if sx.coord_system != sy.coord_system:
         msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
             'have the same coordinate system.'
         raise ValueError(msg.format(sx.name(), sy.name()))
 
-    if sx.coord_system != tx.coord_system and \
-            sx.coord_system is not None and \
-            tx.coord_system is not None:
-        msg = 'The source cube and target grid cube must have the same ' \
-            'coordinate system.'
+    if src_cube.coord_dims(sx) != src_cube.coord_dims(sy):
+        msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
+            'map onto the same cube dimensions.'
+        raise ValueError(msg.format(sx.name(), sy.name()))
+
+    if weights is None:
+        weights = np.ones(sx.shape)
+    if weights.shape != sx.shape:
+        msg = ('Provided weights must have the same shape as the X and Y '
+               'coordinates.')
+        raise ValueError(msg)
+
+    if tx.coord_system != ty.coord_system:
+        msg = 'The target grid cube x ({!r}) and y ({!r}) coordinates must ' \
+            'have the same coordinate system.'
+        raise ValueError(msg.format(tx.name(), ty.name()))
+
+    if sx.coord_system is None or tx.coord_system is None:
```

Comment:
I wonder if it would be worth putting these two coord system checks up just after the checks on source coord system? It would make the flow of these checks more logical (and I almost added a comment asking where the source coord system checks were :stuck_out_tongue_winking_eye: )